### PR TITLE
fix: Remove document-definition.component dependence on line-machine …

### PR DESCRIPTION
**Pull request by pleacu**
…component using a custom event

Fixes: #128

### CODE REVIEW

File: ui/src/app/lib/atlasmap-data-mapper/components/document-definition.component.ts [hunk 1]
```diff
@@ -132,9 +139,16 @@ export class DocumentDefinitionComponent implements OnInit {
...
+  /**
+   * Handle scrolling in this document definition instance.  Avoid a circular dependence with the
+   * LineMachineComponent by dispatching a custom mappings-line-redraw event.
+   * @param event
+   */
   handleScroll(event: any) {
     this.scrollTop = event.target.scrollTop;
-    this.lineMachine.redrawLinesForMappings();
+    this.redrawMappingLinesEvent = new CustomEvent('redrawMappingLines', {detail: this.lineMachine});
+    document.dispatchEvent(this.redrawMappingLinesEvent);
```
Comment (igarashitm) [inline ui/src/app/lib/atlasmap-data-mapper/components/document-definition.component.ts]: oh I just got to know this is a DOM document, built in var in javascript. I'm curious if  it's common way vs. using Angular `EventEmitter`.
https://github.com/syndesisio/syndesis/blob/248a2dbf077f324840b74830f69fe229752ceef1/app/ui/src/app/integration/edit-page/current-flow.service.ts#L352

File: ui/src/app/lib/atlasmap-data-mapper/components/data-mapper-app.component.html [hunk 2]
```diff
@@ -14,13 +14,13 @@
...
         <div style="float:left; width:40%; height:100%;">
-          <document-definition #docDefInput [cfg]="getConfig()" [modalWindow]="modalWindow" [isSource]="true" [lineMachine]="lineMachine"></document-definition>
+          <document-definition #docDefInput [cfg]="getConfig()" [modalWindow]="modalWindow" [isSource]="true"></document-definition>
         </div>
         <div style="float:left; width:20%; height:100%; margin-left:-5px; margin-right:-5px;">
           <line-machine #lineMachine [cfg]="getConfig()" [docDefInput]="docDefInput" [docDefOutput]="docDefOutput"></line-machine>
         </div>
         <div style="float:left; width:40%; height:100%;">
-          <document-definition #docDefOutput [cfg]="getConfig()" [modalWindow]="modalWindow" [isSource]="false" [lineMachine]="lineMachine"></document-definition>
+          <document-definition #docDefOutput [cfg]="getConfig()" [modalWindow]="modalWindow" [isSource]="false"></document-definition>
```
Comment (deeleman) [inline ui/src/app/lib/atlasmap-data-mapper/components/data-mapper-app.component.html]: Id recommend to add a short custom prefix to all new atlasmapper custom elements, so we prevent naming collisions with any other 3rd party library we might implement in the future

File: ui/src/app/lib/atlasmap-data-mapper/components/document-definition.component.ts [hunk 3]
```diff
@@ -61,6 +60,14 @@ export class DocumentDefinitionComponent implements OnInit {
     this.documents = this.cfg.getDocs(this.isSource);
   }
 
+  getLineMachine(): any {
+      return this.lineMachine;
+  }
+
+  setLineMachine(lm: any): void {
+      this.lineMachine = lm;
+  }
```
Comment (deeleman) [inline ui/src/app/lib/atlasmap-data-mapper/components/document-definition.component.ts]: I’d reformat this to two spaces indenting to stay consistent with the rest of the file

File: ui/src/app/lib/atlasmap-data-mapper/components/document-definition.component.ts (hunk 1 above)
Comment (deeleman) [reply ui/src/app/lib/atlasmap-data-mapper/components/document-definition.component.ts]: Exactly, although it’s not commonly used outside the Web Workers arena unless one does not rely on any web components framework such as React, Vue or Angular, which is our case.

I rather stick to one single way to dispatch events in order not to mix up different paradigms in a single architecture and therefore I’d stick to Angulars built in output properties, just for consistency sake. :)

File: ui/src/app/lib/atlasmap-data-mapper/components/line-machine.component.ts [hunk 4]
```diff
@@ -50,13 +50,21 @@ export class LineMachineComponent implements OnInit {
 
   @ViewChild('lineMachineElement') lineMachineElement: ElementRef;
 
-  private yOffset = 3;
+  redrawMappingLinesEventListener = document.addEventListener('redrawMappingLines', function(e: CustomEvent) {
+      const lmcInstance: LineMachineComponent = e.detail;
+      setTimeout(() => {
+          lmcInstance.redrawLinesForMappings();
+          }, 10);
+    }, false);
```
Comment (deeleman) [inline ui/src/app/lib/atlasmap-data-mapper/components/line-machine.component.ts]: We’re mixing native DOM API methods and globals (document, for instance), with callbacks as functions, assignment expressions with typing, parameters and constant names not matching its type, 4 spaces indenting, and defer execution using an arbitrary value (10).

This block might require some love and leverage Angular’s built in API for emitting events instead

File: ui/src/app/lib/atlasmap-data-mapper/components/document-definition.component.ts (hunk 1 above)
Comment (igarashitm) [reply ui/src/app/lib/atlasmap-data-mapper/components/document-definition.component.ts]: And this is an example to emit the event, directly invoke `emit()` on a `CurrentFlowService.events`
https://github.com/syndesisio/syndesis/blob/248a2dbf077f324840b74830f69fe229752ceef1/app/ui/src/app/integration/edit-page/step-configure/data-mapper/data-mapper-host.component.ts#L134

**Pull request closed without merging**